src/dataflow/src/render/mod.rs

```diff
@@ -822,3 +576,450 @@ pub mod datum_vec {
         }
     }
 }
+
+/// An explicit represenation of a rendering plan for provided dataflows.
+pub mod plan {
+
+    use crate::render::join::{DeltaJoinPlan, JoinPlan, LinearJoinPlan};
+    use crate::render::reduce::{KeyValPlan, ReducePlan};
+    use expr::{
+        EvalError, Id, JoinInputMapper, LocalId, MapFilterProject, MirRelationExpr, MirScalarExpr,
+        TableFunc,
+    };
+    use repr::{Datum, Row};
+
+    /// A rendering plan with all conditional logic removed.
+    ///
+    /// This type is exposed publicly but the intent is that its details are under
+    /// the control of this crate, and they are subject to change as we find more
+    /// compelling ways to represent renderable plans. Several stages have already
+    /// encapsulated much of their logic in their own stage-specific plans, and we
+    /// expect more of the plans to do the same in the future, without consultation.
+    #[derive(Debug)]
+    pub enum Plan {
```

Comment:
I think this plan should make explicit what existing indexes are used by the plan and what new indexes are created within. I think that belongs in the planning stage, which would eventually be where our 'cost-based' optimizer would live.

I also think that join planning, as in deciding which strategy the join would use, belongs in this new planning stage, rather than in the transform crate. Planning could receive a flattened join representation and could split the join into subjoins using different strategies depending on the availability of indexes.

If we added information about indexes to this representation of the plan, this would definitely be the output of the optimizer, and hence, should be what we display by default in EXPLAIN.

Now that MFPs are fused, I don't see any reason to keep their Mir counterparts.

Reply:
If we made that explicit here, we could fix #4887 by traversing this plan collecting the indexes actually used.

Reply:
This sounds good. I've mentioned in a comment (https://github.com/MaterializeInc/materialize/pull/7001#issuecomment-858703865)  that I'd like to add this alterwards, as there are a few additional invariants to set up (the stages need to advertise the arrangements they *would* produce, and that implementation needs to be kept in sync with the arrangements that they do form).